ignition/src/IgnitionRobot.cpp

```diff
@@ -945,30 +1134,84 @@ bool IgnitionRobot::setBaseFrame(const gympp::Robot::LinkName& baseLink)
-    basePose.orientation[0] = poseComponent->Data().Rot().W();
-    basePose.orientation[1] = poseComponent->Data().Rot().X();
-    basePose.orientation[2] = poseComponent->Data().Rot().Y();
-    basePose.orientation[3] = poseComponent->Data().Rot().Z();
-
-    return basePose;
+    assert(modelWorldPoseComponent);
+
+    // Extract the transform
+    const ignition::math::Pose3d& world_H_model = modelWorldPoseComponent->Data();
+
+    return Impl::fromIgnitionMath(world_H_model);
 }
 
-gympp::BaseVelocity IgnitionRobot::baseVelocity()
+gympp::Velocity6D IgnitionRobot::baseVelocity()
 {
-    assert(false);
-    return {};
+    // The base velocity is the velocity of the model frame.
+    // Between base link and model frame there's a rigid transformation.
+    // We can extract the velocity of the base link and express it in the model frame.
+
+    // Get the name of the base link
+    const LinkName& baseLink = baseFrame();
+
+    // Get the Pose component of the canonical link.
+    // This is the fixed transformation between the model and the base
+    const auto* canonicalLinkPose =
+        pImpl->ecm->Component<ignition::gazebo::components::Pose>(pImpl->getLinkEntity(baseLink));
+    assert(canonicalLinkPose);
+
+    // Get the transform of the canonical link
+    const ignition::math::Pose3d& model_H_base = canonicalLinkPose->Data();
+
+    // Position of base link wrt model frame
+    const ignition::math::Vector3d& M_o_B_math = model_H_base.Pos();
+    Eigen::Vector3d M_o_B = ignition::math::eigen3::convert(M_o_B_math);
+
+    // Create a skew symmetric matrix from the 3D position
+    Eigen::Matrix3d sk_M_o_B = Eigen::Matrix3d::Zero();
+    sk_M_o_B(1, 0) = M_o_B[2];
+    sk_M_o_B(0, 1) = -M_o_B[2];
+    sk_M_o_B(0, 2) = M_o_B[1];
+    sk_M_o_B(2, 0) = -M_o_B[1];
+    sk_M_o_B(2, 1) = M_o_B[0];
+    sk_M_o_B(1, 2) = -M_o_B[0];
+
+    // Get the rotation of the base link frame wrt the model frame
+    ignition::math::Matrix3d M_R_B_math(model_H_base.Rot());
+    Eigen::Matrix3d M_R_B = ignition::math::eigen3::convert(M_R_B_math);
+
+    // Construct the velocity transformation matrix
+    Eigen::Matrix<double, 6, 6> M_X_B = Eigen::Matrix<double, 6, 6>::Zero();
+    M_X_B.topLeftCorner(3, 3) = M_R_B;
+    M_X_B.bottomRightCorner(3, 3) = M_R_B;
+    M_X_B.topRightCorner(3, 3) = sk_M_o_B * M_R_B;
+
+    // Get the velocity of the base link
+    gympp::Velocity6D baseVelocity = linkVelocity(baseLink);
+    Eigen::Matrix<double, 6, 1> baseVelocityEigen;
+    baseVelocityEigen[0] = baseVelocity.linear[0];
+    baseVelocityEigen[1] = baseVelocity.linear[1];
+    baseVelocityEigen[2] = baseVelocity.linear[2];
+    baseVelocityEigen[3] = baseVelocity.angular[0];
+    baseVelocityEigen[4] = baseVelocity.angular[1];
+    baseVelocityEigen[5] = baseVelocity.angular[2];
+
+    // Express the velocity in the model frame
+    Eigen::Matrix<double, 6, 1> modelVelocityEigen = M_X_B * baseVelocityEigen;
```

Comment:
Why don't you simply use Ignition Math or  Eigen methods for the cross product, such as: 
~~~
ignition::math::Vector3d modelLinVel = M_R_B*baseLinVel + M_o_B_math.Cross(M_R_B*baseAngVel)
ignition::math::Vector3d modelAngVel = M_R_B*baseLinVel;
~~~
or something similar? Writing down explicitly the skew simmetric matrix is quite error prone.

Furthermore, how did you derive the math for this transform? If the linear and  angular velocity are expressed in the world frame, and is what we typically refer as "mixed" representation, there is no reason to multiply for the rotation matrix. For a recap of  all the transformation rules between different velocity representation, see Table 2.1 at Page 20 of https://traversaro.github.io/phd-thesis/traversaro-phd-thesis.pdf (code available at https://github.com/traversaro/phd-thesis/blob/143bf6e070006e74fce6ea0b9ed5abd89bee2d4f/thesis/Chapter20RigidBody/chapter-rigidbody.tex#L657).

Reply:
Changed as explained in https://github.com/robotology/gym-ignition/pull/117#discussion_r363124317

Reply:
Cool, I was not aware of the `Cross` method! Though, accordingly to https://github.com/robotology/gym-ignition/pull/117#discussion_r363124317, this transformation is not necessary.